client/src/main/scala/org/http4s/armeria/client/ArmeriaClient.scala

```diff
@@ -103,17 +103,21 @@ private[armeria] final class ArmeriaClient[F[_]] private[client] (
     } yield Response(status = status, headers = toHeaders(headers), body = body)
   }
 
-  /** Converts [[java.util.concurrent.CompletableFuture]] to `F[_]` */
-  private def fromCompletableFuture(
-      completableFuture: CompletableFuture[ResponseHeaders]): F[ResponseHeaders] =
-    F.async_[ResponseHeaders] { cb =>
-      val _ = completableFuture.handle { (result, ex) =>
-        if (ex != null)
-          cb(Left(ex))
-        else
-          cb(Right(result))
-        null
+  private def fromCompletableFuture(cf: => CompletableFuture[ResponseHeaders]): F[ResponseHeaders] =
+    F.async { cb =>
+      cf.handle[Unit] { (result: ResponseHeaders, err: Throwable) =>
+        err match {
+          case null =>
+            cb(Right(result))
+          case _: CancellationException =>
+            ()
+          case ex: CompletionException if ex.getCause ne null =>
+            cb(Left(ex.getCause))
+          case ex =>
+            cb(Left(ex))
+        }
       }
+      F.delay(Some(F.delay(cf.cancel(true)).void))
```

Comment:
If we log this:

```suggestion
      F.delay(Some(F.delay(println("canceled: " + cf.cancel(true))).void))
```

we discover that it didn't actually cancel:

```
canceled: false
org.http4s.armeria.client.ArmeriaClientSuite:
  + timeout response 0.926s
```

So it is technically a leak to pretend like it did. However, you might decide this is the right compromise for this situation.

Reply:
`splitResponse.headers` returns `HeadersFuture` which is an `UnmodifiedFuture` used to deliver a completion signal only one way.
https://github.com/line/armeria/blob/f0818b990b47ef863bf82bad81f34c09eb2ad565/core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpResponse.java#L57
https://github.com/line/armeria/blob/cae3b6da373101c0976e2d00f1f3cbb706d07e00/core/src/main/java/com/linecorp/armeria/common/util/UnmodifiableFuture.java#L182-L188

Armeria uses `HttpRequest.abort()` or `HttpResponse.abort()` to cancel a request process and stop receiving new messages from the network or services.